site/content/blog/2020-01-08-incorporating-golangci-lint.md

```diff
@@ -0,0 +1,140 @@
+
+We needed a way to run all of these linters together and efficiently too. Gometalinter was a pioneer in this field by providing a way to run multiple linters together. Unfortunately, it did this by shelling out to the linters and running them, which made it very inefficient. To its defence, using all the linters as a library was also equally hard, because there wasn’t any standard API that all of them could use.
+
+But the introduction of the `golang.org/x/tools/go/analysis` package during the 1.12 cycle was a game-changer. It introduced a standard API for writing Go static analyzers, which allowed them to be easily shared with the rest of the ecosystem in a plug-and-play model. Finally the One Ring to rule them all!
+
+Unsurprisingly, not before too long, GolangCI-Lint came up on the horizon. It used the go/analysis API to load all the linters and ran them concurrently, leading to a drastic reduction in memory and improvement in speed; becoming the successor to Gometalinter which is deprecated now.
+
+The writing was on the wall. GolangCI-Lint was the right choice. Following is an account of how we integrated it into our codebase.
+
+
+#### Integrating GolangCI-Lint into our CI
+
+Right off the bat, there was a big question mark hanging over our heads. It would be very, _very_ painful to have to fix _all_ the issues in our codebase before enforcing it as a CI check, by which time, new issues would have creeped in. It would be a never-ending chase.
+
+Fortunately, GolangCI-Lint provides us with a `--new-from-rev=HEAD~1` option which allowed it to check only code added in the latest commit. This gives us breathing room to enable the check in our CI and retroactively fix all the old issues. Granted, that it won’t catch issues for PRs with more than one commit, but it’s a start and we planned to lift this restriction anyways.
+
+We also decided to start off with the Github [check](https://golangci.com/) which relieves any CI setup burden on our side. This check helps community members spot linter issues at first glance.
+
+#### Deciding the initial set of linters
+
+So with that taken care of, our next task was to decide on a set of linters to start off with. We could not begin with a huge set because first of all, it will take longer to run. And secondly, it would further delay our target of fixing all the issues already existing in our codebase. We needed a small but good enough set of linters that would catch effective issues, but wouldn’t take too long to fix.
```

Comment:
```suggestion
So with that taken care of, our next task was to decide on a set of linters to start off with. We couldn't begin with a huge set because first of all, it would take longer to run. Additionally, it would further delay our target of fixing all the  existing issues in our codebase. We needed a small but powerful set of linters that would catch effective issues, but wouldn’t take too long to fix.
```